homeassistant/components/recorder/websocket_api.py

```diff
@@ -106,3 +111,38 @@ def ws_info(
         "thread_running": thread_alive,
     }
     connection.send_result(msg["id"], recorder_info)
+
+
+@websocket_api.require_admin
+@websocket_api.websocket_command({vol.Required("type"): "backup/start"})
+@websocket_api.async_response
+async def ws_backup_start(
+    hass: HomeAssistant, connection: websocket_api.ActiveConnection, msg: dict
+) -> None:
+    """Backup start notification."""
+
+    _LOGGER.info("Backup start notification, locking database for writes")
+    instance: Recorder = hass.data[DATA_INSTANCE]
+    try:
+        await instance.lock_database()
+    except TimeoutError as err:
+        connection.send_error(msg["id"], "timeout_error", str(err))
+        return
+    connection.send_result(msg["id"])
+
+
+@websocket_api.require_admin
+@websocket_api.websocket_command({vol.Required("type"): "backup/end"})
+@websocket_api.async_response
+async def ws_backup_end(
```

Comment:
I know I am a little late to the party, but I thought of some improvements:

- Backup logic should live in the `hassio` integration. Recorder (and potentially in the future others) should indicate they have interest in this. Maybe via a `hassio.py` platform.
- WS API should be in hassio too.

And here for a bit more crazier one:
- If the supervisor were to disconnect abruptly, it would never call `backup_end`. That will leave the system in a broken state. It would be better if the backup_start would store the "end backup" logic in [`connection.subscriptions`](https://github.com/home-assistant/core/blob/dev/homeassistant/components/websocket_api/commands.py#L114-L118). Subscriptions are automatically finished when the connection is disconnected. That way, if the supervisor's websocket connection breaks, the backup mode is ended. The supervisor can then [end the subscription](https://developers.home-assistant.io/docs/api/websocket#unsubscribing-from-events) when it's done to also end the in-backup mode.

Reply:
> * Backup logic should live in the hassio integration. Recorder (and potentially in the future others) should indicate they have interest in this. Maybe via a hassio.py platform.
> * WS API should be in hassio too.

Agreed, also suggested that in the PR intro (at least the WS part). At that point there was no other user, and I got no push back :angel: 

> If the supervisor were to disconnect abruptly, it would never call backup_end. That will leave the system in a broken state.

It would queue up events up to `MAX_QUEUE_BACKLOG * 0.9`, and then release the lock.

> It would be better if the backup_start would store the "end backup" logic in connection.subscriptions. 

Wouldn't that mean a single long running WS call from Supervisor side? That doesn't work, as I need to wait until the database is really locked (so `backup_start` not only notifies Core that backup is about to start, its return also signals the Supervisor that the Core is in fact ready to get going...).

Reply:
A subscription can send data over it to the caller. 

The Supervisor keeps it's WS API open the whole time I think.

Reply:
> Backup logic should live in the hassio integration. Recorder (and potentially in the future others) should indicate they have interest in this. Maybe via a hassio.py platform.

This seems backwards IMHO, the recorder does not care about hassio?